network/p2p/topic_validator.go

```diff
@@ -0,0 +1,31 @@
+package p2p
+
+import (
+	"context"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type StakedValidator struct {
+	stakedIdentities func() flow.IdentityList
+}
+
+func (v *StakedValidator) Validate(ctx context.Context, receivedFrom peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
+	// check that message contains a valid sender ID
+	if from, err := messageSigningID(msg); err == nil {
+		// check that the sender peer ID matches a staked Flow key
+		for _, id := range v.stakedIdentities() {
+			if key, err := LibP2PPublicKeyFromFlow(id.NetworkPubKey); err == nil {
+				if pid, err := peer.IDFromPublicKey(key); err == nil {
+					if from == pid {
+						return pubsub.ValidationAccept
+					}
+				}
+			}
```

Comment:
Could we use `PeerIDFromFlowPublicKey` ?

Also, I'm kinda mifed we recompute this on every message. I understand the idea is for the closure in `stakedIdentities` to control the refresh of identities, but it would be great to fetch the libp2p identities from a cache if `stakedIdentities()` returns the same as its cached value.

Reply:
Ah, I see the import cycle ...